vcd/resource_vcd_nsxt_edgegateway.go

```diff
@@ -329,9 +417,147 @@ func getNsxtEdgeGatewayType(d *schema.ResourceData, vcdClient *VCDClient, isCrea
+
+	// 'auto_subnet' and 'total_allocated_ip_count' are specified
+	case usingAutoSubnetAllocation:
+		isPrimaryIpSet, subnetValues := getNsxtEdgeGatewayUplinksTypeAutoSubnets(d)
+
+		uplinks := []types.EdgeGatewayUplinks{{}}
+		uplinks[0].Subnets = types.OpenAPIEdgeGatewaySubnets{Values: subnetValues}
+		if totalIpCount, isSetTotalIpCount := d.GetOk("total_allocated_ip_count"); isSetTotalIpCount {
+
+			// For create operation, the total_allocated_ip_count cannot be negative and therefore we can utilize QuickAddAllocatedIPCount field
+			uplinks[0].QuickAddAllocatedIPCount = totalIpCount.(int)
+			// Primary IP is an additional IP address that is not included in the total allocated IP
+			// count when used with QuickAddAllocatedIPCount therefore we need to subtract it
+			if isPrimaryIpSet {
+				uplinks[0].QuickAddAllocatedIPCount = totalIpCount.(int) - 1
+			}
+
+		}
+		return uplinks, nil
+
+	// 'auto_allocated_subnet' is specified
+	case usingAutoAllocatedSubnetAllocation:
+		return []types.EdgeGatewayUplinks{{Subnets: types.OpenAPIEdgeGatewaySubnets{Values: getNsxtEdgeGatewayUplinksTypeAutoAllocateSubnets(d)}}}, nil
+
+	}
+
+	return nil, fmt.Errorf("one of the following fields must be set: 'subnet', 'auto_subnet', 'auto_allocated_subnet'")
+}
+
+// getNsxtEdgeGatewayUplinksTypeForUpdate handles uplink structure in update only operations
+func getNsxtEdgeGatewayUplinksTypeForUpdate(d *schema.ResourceData, currentlyAllocatedIpCount *int, edgeGateway *govcd.NsxtEdgeGateway) ([]types.EdgeGatewayUplinks, error) {
+	if edgeGateway == nil {
+		return nil, fmt.Errorf("edge gateway cannot be nil")
+	}
+
+	if currentlyAllocatedIpCount == nil {
+		return nil, fmt.Errorf("currentlyAllocatedIpCount cannot be nil for update operation")
+	}
+
+	_, usingSubnetAllocation := d.GetOk("subnet")
+	_, usingAutoSubnetAllocation := d.GetOk("auto_subnet")
+	_, usingAutoAllocatedSubnetAllocation := d.GetOk("auto_allocated_subnet")
+
+	log.Printf("[TRACE] NSX-T Edge Gateway update 'subnet': %t (HasChange %t), 'auto_subnet': %t (HasChange %t), 'auto_allocated_subnet': %t (HasChange %t)",
+		usingSubnetAllocation, d.HasChange("subnet"), usingAutoSubnetAllocation, d.HasChange("auto_subnet"), usingAutoAllocatedSubnetAllocation, d.HasChange("auto_allocated_subnet"))
+
+	switch {
+	// 'subnet' is specified
+	case d.HasChange("subnet"):
+		return []types.EdgeGatewayUplinks{{Subnets: types.OpenAPIEdgeGatewaySubnets{Values: getNsxtEdgeGatewayUplinksType(d)}}}, nil
+
+	// 'auto_subnet' and 'total_allocated_ip_count' are specified
+	case d.HasChange("auto_subnet") || d.HasChange("total_allocated_ip_count"):
+		_, subnetValues := getNsxtEdgeGatewayUplinksTypeAutoSubnets(d)
+		uplinks := []types.EdgeGatewayUplinks{{}}
+		uplinks[0].Subnets = types.OpenAPIEdgeGatewaySubnets{Values: subnetValues}
+		if desiredtotalIpCount, isSetTotalIpCount := d.GetOk("total_allocated_ip_count"); isSetTotalIpCount {
+			if desiredtotalIpCount.(int) < 1 {
```

Comment:
I was wondering, whether we could use
```
ValidateFunc: validation.IntAtLeast(0),
```
in `total_allocated_ip_count` attribute definition inside the schema?

Reply:
I have changed the code to have `ValidateFunc: validation.IntAtLeast(1),` (it must be at least 1 due to primary_ip)